usr/lib/linuxmint/mintSources/mintSources.py

```diff
@@ -1141,48 +1166,58 @@ def add_ppa(self, widget):
-            info_text = "%s\n\n%s\n\n%s\n\n%s" % (line, self.format_string(ppa_info["displayname"]), self.format_string(ppa_info["description"]), str(ppa_info["web_link"]))
+            info_text = "%s\n\n%s\n\n%s\n\n%s" % (line,
+                self.format_string(ppa_info["displayname"]),
+                self.format_string(ppa_info["description"]), str(ppa_info["web_link"]))
             if self.show_confirm_ppa_dialog(self._main_window, info_text):
                 (deb_line, file) = expand_ppa_line(line.strip(), self.config["general"]["base_codename"])
                 deb_line = expand_http_line(deb_line, self.config["general"]["base_codename"])
                 debsrc_line = 'deb-src' + deb_line[3:]
 
-                # Add the key
-                subprocess.call(["apt-key", "adv", "--keyserver", "keyserver.ubuntu.com", "--recv-keys", ppa_info["signing_key_fingerprint"]])
-                self.load_keys()
+                # Add the key if not in keyring
+                self.add_new_key(ppa_info["signing_key_fingerprint"])
 
                 # Add the PPA in sources.list.d
+                sources_enabled = self.builder.get_object("source_code_switch").get_active()
                 with open(file, "w", encoding="utf-8", errors="ignore") as text_file:
                     text_file.write("%s\n" % deb_line)
-                    text_file.write("%s\n" % "# "+debsrc_line)
-
-                # Add the package line in the UI
-                repository = Repository(self, deb_line, file, True)
-                self.ppas.append(repository)
-                tree_iter = self._ppa_model.append((repository, repository.selected, repository.get_ppa_name()))
-
-                # Add the source line in the UI
-                repository = Repository(self, debsrc_line, file, False)
-                self.ppas.append(repository)
-                tree_iter = self._ppa_model.append((repository, repository.selected, repository.get_ppa_name()))
+                    text_file.write("%s%s\n" % ("" if sources_enabled else "# ", debsrc_line))
+
+                # Add the package line to the UI or replace it if it exists
+                def add_to_ui(line):
+                    repo = next((repo for repo in self.ppas if (repo.line == line and repo.file == file)), None)
+                    if repo:
+                        iter = next((item.iter for item in self._ppa_model if line in self._ppa_model.get_value(item.iter, 2).split("\n")), None)
+                        if iter:
+                            self._ppa_model.remove(iter)
+                        self.ppas.remove(repo)
+                    repository = Repository(self, line, file, True)
+                    self.ppas.append(repository)
+                    tree_iter = self._ppa_model.append((repository, repository.selected, repository.get_ppa_name()))
+
+                add_to_ui(deb_line)
+                add_to_ui(debsrc_line)
```

Comment:
Wouldn't it be better to just bluntly refuse to re-add a PPA that we already have? just like we throw "Invalid PPA name", or "PPA not found", we could throw a "PPA already exists or whatever". It would be more explicit for the user and it would simplify things here.. no need to find/replace.